// multicraft/points/users.js
module.exports = api => {
  return {
    listUsers: async () => {
      return await api.send("listUsers");
    },

    findUsers: async (field, value) => {
      return await api.send(
        "findUsers",
        `"${field}""${value}"`,
        { field, value },
        true
      );
    },

    getUser: async id => {
      return await api.send("getUser", id, { id });
    },

    getCurrentUser: async () => {
      return await api.send("getCurrentUser");
    },

    updateUser: async (id, field, value) => {
      return await api.send(
        "updateUser",
        `"${field}""${value}"`,
        { field, value },
        true
      );
    },

    validateUser: async (name, password) => {
      return await api.send("validateUser", `${name}${password}`, {
        name,
        password
      });
    },
    
    createUser: async (name, email, password) => {
      return await api.send(
        "createUser",
        `"${name}""${email}""${password}"`,
        { name, email, password },
        true
      );
    },
    
    deleteUser: async id => {
      return await api.send("deleteUser", id, {id});
    },
    
    getUserApiKey: async user_id => {
      return await api.send("getUserApiKey", user_id, {user_id});
    },
    
    generateUserApiKey: async user_id => {
      return await api.send("generateUserApiKey", user_id, {user_id});
    },
    
    getUserId: async name => {
      return await api.send("getUserId", name, {name});
    },
    
  };
};
